keps/sig-node/4742-node-labels-downward/README.md

```diff
@@ -0,0 +1,809 @@
+example is utilizing topology data from cloud providers, which are automatically
+assignedto nodes, and then used to tailor workloads to minimize cross-zone data
+transfers and reduce costs. Having an easy way for pods to access these node
+topology labels would provide users a straightforward, maintainable way to
+optimize their workloads given topology constraints.
+
+Workarounds today typically involve using an initContainer to query the
+Kubernetes API and then pass data via shared volume to other containers within
+the same pod. This adds additional demand on the API server and is burdensome
+compared to the ease of using downwardAPI for pod labels and metadata.
+
+### Goals
+
+* Gain access to node labels in form of `topology.k8s.io/*` on pods through volume mounts
+* Gain access to node labels in form of `topology.k8s.io/*` on pods through environmental variables
+
+### Non-Goals
+
+* Not to expose additional node info outside of labels
+* Not to pass any additional node labels other than `topology.k8s.io/*` to pods
+* Not to guarantee the label value assigned at pod creation is the most recent node label value because it is assigned at pod creation time
+
+## Proposal
+
+The initial design includes:
+
+* Being able to pass node labels to volumes with fieldPath of `node.metadata.labels` where only labels containing `topology.k8s.io/` are passed through
+* Being able to pass a specific node label through an env vars like `topology.k8s.io/region`
+
+### User Stories
+
+* As a cluster operator, I want to make decisions based on node topology labels.
+* As a cluster operator, I want to access node topology labels inside of my pod
+* As a cluster operator, I want to access node instance types labels inside of my pod
+* As a developer, I want to know which region my app is serving, to be able to diagnose problems they may face in certain AZs or regions
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+* Scope creep. Allowing additional node information or node label info could
+create security issues. This is mitigated by limiting the node information
+available to the pod to strictly node labels, and specifically those of the
+form `topology.k8s.io/*`.
+
+* Exposing sensitive data as node labels to pods. This is mitigated by ensuring
+node labels contain the specific pattern `topology.k8s.io/*` in order to be
+available to Pods.
+
+* Stale data. Information obtained through node labels is like information
+attained through a configmap or secret mounted to a pod, being passed on
+creation but not guaranteed to be immutable and thus should be treated as so.
+
+* Can potentially be featuregated to make this feature strictly opt-in, if
```

Comment:
feature gates are not long-term flags -- if you need a mechanism for admisn to disable this entirely, that's not a feature gate (and I would argue it indicates a deeper problem)